File: Data Structures & Algorithms/max-water-container/submission-1.js
class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     */

    maxArea(heights) {
        let left = 0;
        let right = heights.length - 1;
        let maxArea = 0;
        while(left < right)
        {
            let bottleneck = Math.min(heights[left], heights[right]);
            let area = bottleneck * (right - left);
            maxArea = Math.max(maxArea, area);

            // move the pointer of the smaller height
            if(heights[left] < heights[right]){
                left = left + 1;
            }
            else{
                right = right - 1;
            }
        }
        return maxArea;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.cs
public class Solution {
    public int CharacterReplacement(string s, int k) {
        // stupid
        int[] counts = new int[128]; 
        int left = 0;
        int maxLen = 0;
        int maxFreq = 0;

        for (int right = 0; right < s.Length; right++) {
            // add the new character to our windows count
            counts[s[right]]++;
            
            // update our record for the highest frequency seen in the window
            maxFreq = Math.Max(maxFreq, counts[s[right]]);

            // scam. window length - most frequent character count
            int windowLen = right - left + 1;
            if (windowLen - maxFreq > k) {
                // if it costs too many replacements, the window is invalid.
                // we must shrink it from the left. 
                
                // FIRST, remove the character at the left pointer from our counts
                counts[s[left]]--;
                // THEN move the left pointer
                left++; 
            }
            maxLen = Math.Max(maxLen, right - left + 1);
        }

        return maxLen;
    }
}